File: spring.py
# uncomment the next line if running in a notebook
# %matplotlib inline
import numpy as np
import time
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
start_time = time.time()
# mass, spring constant, initial position and velocity
m = 1
k = 1
x = 0
v = 1

# simulation time, time-step and time
t_max = 100
dt = 0.01
t_array = np.arange(0, t_max, dt)

# initialise empty lists to record trajectories
x_list = []
v_list = []
# Euler integration

x_list_vert = []
v_list_vert = []
m_vert = 1
k_vert = 1
x_vert = 0
v_vert = 1
for t in t_array:

    # append current state to trajectories
    x_list.append(x)
    v_list.append(v)

    # calculate new position and velocity
    a = -k * x / m
    x = x + dt * v
    v = v + dt * a
for t in t_array:
    # Vertlet Integration
    if t == 0:
        x_previous = x_vert;
        x1 = x_vert + dt*v_vert;
        x_list_vert.append(x1)
        v_list_vert.append(v_vert)
    if t !=0:
        x_current = x1
        x1 = x_current*2 - x_previous -((dt*dt)*k*x_current / m)
        x_previous = x_current
        v_vert = (1/(2*dt))*(x1-x_previous);
        x_list_vert.append(x1)
        v_list_vert.append(v_vert)

# convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
x_array = np.array(x_list)
v_array = np.array(v_list)

x_vert_array = np.array(x_list_vert)
v_vert_array = np.array(v_list_vert)
end_time = time.time()
time = end_time - start_time
print("Elapsed Time: " + str(time))
# plot the position-time graph
plt.figure(1)
plt.clf()
plt.xlabel('Euler Integration: time (s)')
plt.grid()
plt.plot(t_array, x_array, label='x (m)')
plt.plot(t_array, v_array, label='v (m/s)')
plt.legend()
plt.show()

plt.figure(2)
plt.clf()
plt.xlabel('Vertlet Integration: time (s)')
plt.grid()
plt.plot(t_array, x_vert_array, label='x (m)')
plt.plot(t_array, v_vert_array, label='v (m/s)')
plt.legend()
plt.show()

File: spring.cpp
#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

// t  =  100, dt = 0.00001
// Execution time for Python code with Vertlet Integration is: 17.146 seconds
// Execution time for C++ code with Vertlet Integration and no optimization is: 1.561 seconds
// Execution time for C++ code with Vertlet Integration and fast optimization is: 0.445 seconds
// Execution time for C++ code with Vertlet Integration and faster optimization is: 0.399 seconds
// Execution time for C++ code with Vertlet Integration and fastest optimization is: 0.392 seconds
// Execution time for C++ code with Vertlet Integration and fastest, smallest optimization is: 0.399 seconds
// Execution time for C++ code with Vertlet Integration and fastest, agressive optimization is: 0.393 seconds



int main() {
    clock_t start, end;
    start = clock();
    // declare variables
    double m, k, x, v, t_max, dt, t,x_previous,x1,x_current;
    vector<double> t_list, x_list, v_list;
    
    // mass, spring constant, initial position and velocity
    m = 1;
    k = 1;
    x = 0;
    v = 1;
    
    // simulation time and timestep
    t_max = 100;
    dt = 0.00001;
    
    // Vertlet integration
    for (t = 0; t <= t_max; t = t + dt) {
        if(t == 0){
            x_previous = x;
            x1 = x + dt*v;
        }
        // append current state to trajectories
        x_current = x1;
        x1 = x_current*2 - x_previous -((dt*dt)*k*x_current / m);
        x_previous = x_current;
        v = (1/(2*dt))*(x1-x_previous);
        t_list.push_back(t);
        x_list.push_back(x1);
        v_list.push_back(v);
    }
    // Write the trajectories to file
    ofstream fout;
    fout.open("trajectories.txt");
    if (fout) { // file opened successfully
        for (int i = 0; i < t_list.size(); i = i + 1) {
            fout << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
        }
    } else { // file did not open successfully
        cout << "Could not open trajectory file for writing" << endl;
    }
    end = clock();
    double time_taken = double(end-start) / double(CLOCKS_PER_SEC);
    cout << "Time taken by program is : " << fixed
    << time_taken;
    cout << " sec " << endl;
    /* The file can be loaded and visualised in Python as follows:
     
     import numpy as np
     import matplotlib.pyplot as plt
     results = np.loadtxt('trajectories.txt')
     plt.figure(1)
     plt.clf()
     plt.xlabel('time (s)')
     plt.grid()
     plt.plot(results[:, 0], results[:, 1], label='x (m)')
     plt.plot(results[:, 0], results[:, 2], label='v (m/s)')
     plt.legend()
     plt.show()
     
     */
}
